drv/cosmo-hf/src/main.rs

```diff
@@ -177,13 +162,47 @@ impl FlashDriver {
         });
     }
 
+    /// Wait for a condition represented by the provided `poll` function.
+    ///
+    /// The driver will wait until the `poll` function returns `true`. Each time
+    /// `poll` returns `false`, the provided `trace` will be recorded in the
+    /// ring buffer.
+    fn poll_wait(&self, poll: fn(&Self) -> bool, trace: Trace) {
```

Comment:
Have you looked at whether this gets inlined in practice?

If it does, you could make the function more general with no added code size implications by loosening the `poll` type to `impl FnMut(&self) -> bool`. That'd allow closures with mutable state.

If the function is not currently inlined then this would cause it to be duplicated, potentially at a small code size cost. Personally, I'd probably still do it unless it is clearly a serious problem, because taking general purpose closures is nice.

Reply:
Ah, yeah, good point. I used the function pointer as I wanted to avoid monomorphization in case it was *not* being inlined, and none of the _current_ code needed `FnMut`. But, it occurs to me that we could also mark the function as `#[inline(always)]`...